backend/controller/dal/fsm.go

```diff
@@ -79,11 +84,17 @@ func (d *DAL) StartFSMTransition(ctx context.Context, fsm schema.RefKey, executi
 	return nil
 }
 
-func (d *DAL) FinishFSMTransition(ctx context.Context, fsm schema.RefKey, instanceKey string) error {
-	_, err := d.db.FinishFSMTransition(ctx, fsm, instanceKey)
-	observability.FSM.TransitionCompleted(ctx, fsm)
-
-	return dalerrs.TranslatePGError(err)
+// FinishFSMTransition marks an FSM transition as completed and updates the instance struct.
```

Comment:
Let's return an updated clone of the struct instead of mutating it in place; mutable parameters are not very obvious, and can lead to some very surprising outcomes.